Source/DafnyCore/Compilers/Rust/Dafny-compiler-rust.dfy

```diff
@@ -1,126 +1,508 @@
+      content
+    }
+  }
+
+  datatype Trait =
+    | Trait(typeParams: seq<TypeParam>, tpe: Type, where: string, body: seq<ImplMember>)
+  {
+    function ToString(ind: string): string {
+      "trait " + TypeParam.ToStringMultiple(typeParams, ind) + tpe.ToString(ind)
+      + (if where != "" then "\n" + ind + IND + where else "")
+      + " {" +
+      SeqToString(body, (member: ImplMember) => "\n" + ind + IND + member.ToString(ind + IND), "")
+      + (if |body| == 0 then "" else "\n" + ind) + "}"
+    }
+  }
+
+  datatype Impl =
+    | ImplFor(typeParams: seq<TypeParam>, tpe: Type, forType: Type, where: string, body: seq<ImplMember>)
+    | Impl(typeParams: seq<TypeParam>, tpe: Type, where: string, body: seq<ImplMember>)
+  {
+    function ToString(ind: string): string {
+      "impl " + TypeParam.ToStringMultiple(typeParams, ind) + tpe.ToString(ind)
+      + (if ImplFor? then " for " + forType.ToString(ind + IND) else "")
+      + (if where != "" then "\n" + ind + IND + where else "")
+      + " {" +
+      SeqToString(body, (member: ImplMember) => "\n" + ind + IND + member.ToString(ind + IND), "")
+      + (if |body| == 0 then "" else "\n" + ind) + "}"
+    }
+  }
+  datatype ImplMember =
+    | RawImplMember(content: string)
+    | FnDecl(pub: VISIBILITY, fun: Fn)
+  {
+    function ToString(ind: string): string {
+      if FnDecl? then
+        (if pub == PUB then "pub " else "") + fun.ToString(ind)
+      else assert RawImplMember?; content
+    }
+  }
+  newtype VISIBILITY = x: int | 0 <= x < 2
+  const PUB := 1 as VISIBILITY
+  const PRIV := 0 as VISIBILITY
+
+  datatype Formal =
+    Formal(name: string, tpe: Type)
+  {
+    function ToString(ind: string): string {
+      if name == "self" && tpe.SelfOwned? then name
+      else if name == "&self" && tpe.Self? then name
+      else if name == "&mut self" && tpe.SelfMut? then name
+      else
+        name + ": " + tpe.ToString(ind)
+    }
+    static const self := Formal("&self", Self)
+    static const selfOwned := Formal("self", SelfOwned)
+    static const selfMut := Formal("&mut self", SelfMut)
+  }
+
+  datatype Pattern =
+    RawPattern(content: string)
+  {
+    function ToString(ind: string): string {
+      content
+    }
+  }
+
+  datatype MatchCase =
+    MatchCase(pattern: Pattern, rhs: Expr)
+  {
+    function ToString(ind: string): string {
+      var newIndent := if rhs.Block? then ind else ind + IND;
+      var rhsString := rhs.ToString(newIndent);
+
+      pattern.ToString(ind) + " =>" +
+      if '\n' in rhsString && rhsString[0] != '{' then "\n" + ind + IND + rhsString
+      else " " + rhsString
+    }
+  }
+
+  datatype AssignIdentifier =
+    AssignIdentifier(identifier: string, rhs: Expr)
+  {
+    function ToString(ind: string): string {
+      identifier + ": " + rhs.ToString(ind + IND)
+    }
+  }
+
+  // When a raw stream is given, try to put some indentation on it
+  function AddIndent(raw: string, ind: string): string {
```

Comment:
Yep, although I might actually use this one later on if I decide that the indentation I asked for is not enough, i.e. because the string is too long and I want to add a newline, I'd add some indentation.